File: cpp/src/event_model/transaction_event_v1.cpp
#include "event_model/transaction_event_v1.h"
#include <stdexcept>
#include <algorithm>

namespace event_model {
    TransactionEventV1::TransactionEventV1(
        Uuid account_id,
        std::optional<std::string> external_transaction_id,
        std::int64_t amount_minor,
        std::int32_t precision,
        std::array<char, 3> currency,
        std::optional<std::string> raw_description,
        Timestamp occurred_at,
        Uuid event_id,
        SourceType source,
        std::optional<std::string> source_connector,
        std::optional<std::string> source_institution,
        Timestamp ingested_at,
        std::optional<Uuid> batch_id,
        std::optional<std::array<char, 64> > source_file_hash_hex
    )
        : account_id_(account_id)
          , external_transaction_id_(std::move(external_transaction_id))
          , amount_minor_(amount_minor)
          , precision_(precision)
          , currency_(currency)
          , raw_description_(std::move(raw_description))
          , occurred_at_(occurred_at)
          , event_id_(event_id)
          , schema_version_(1)
          , source_(source)
          , source_connector_(std::move(source_connector))
          , source_institution_(std::move(source_institution))
          , ingested_at_(ingested_at)
          , batch_id_(batch_id)
          , source_file_hash_hex_(source_file_hash_hex) {
        if (occurred_at_ > ingested_at_) {
            throw std::invalid_argument("occurred_at must be <= ingested at");
        }

        for (char c: currency_) {
            if (c < 'A' || c > 'Z') {
                throw std::invalid_argument("currency must be uppercase ISO-4217");
            }
        }

        if (precision_ < 0) {
            throw std::invalid_argument("precision must be non-negative");
        }
    }
    const Uuid& TransactionEventV1::account_id() const noexcept {
        return account_id_;
    }
    std::int64_t TransactionEventV1::amount_minor() const noexcept {
        return amount_minor_;
    }
    const std::array<char, 3>& TransactionEventV1::currency() const noexcept {
        return currency_;
    }
    Timestamp TransactionEventV1::occurred_at() const noexcept {
        return occurred_at_;
    }
    Timestamp TransactionEventV1::ingested_at() const noexcept {
        return ingested_at_;
    }
    SourceType TransactionEventV1::source() const noexcept {
        return source_;
    }
}

File: python/event_model/transaction_event_v1.py
#Produce a TransactionEventV1DTO object (or dict) that is guaranteed well-formed

import json
import sys

file_path = 'data.json'

# list of spec tuples as the source of truth in deterministic order matching the V1 schema (field, expected_type, required)

validation_fields = [
    # Domain Payload
    ('account_id', str, True),
    ('external_transaction_id', str, False),
    ('amount_minor', int, True),
    ('precision', int, True),
    ('currency', str, True),
    ('raw_description', str, False),
    ('occurred_at', str, True),
    # Log Metadata
    ('event_id', str, True),
    ('schema_version', int, True),
    ('source', str, True),
    ('source_connector', str, False),
    ('source_institution', str, False),
    ('ingested_at', str, True),
    ('batch_id', str, False),
    ('source_file_hash', str, False)
]

try:
    with open(file_path, 'r') as file:
        data = json.load(file)

    if not isinstance(data, dict):
        raise ValueError(f"{type(data)} is not in hash/dictionary format.")

    error_list = []
    event_payload = {}

    for name, expected_type, required in validation_fields:
        if required:
            if name not in data:
                error_list.append(name + ": required field missing")
                continue
            elif data[name] is None:
                error_list.append(name + ": required field is null")
                continue
        else:
            if name not in data or data[name] is None:
                event_payload[name] = None
                continue
        # boolean masquerading as int check
        if expected_type is int and isinstance(data[name], bool):
            error_list.append(name + ": invalid value type")
        elif not isinstance(data[name], expected_type):
            error_list.append(name + ": invalid value type")
        else:
            event_payload[name] = data[name]

    # initial error check to stop execution if any fields are missing or wrong type at runtime
    if error_list:
        raise ValueError(f"Schema Validation Failed\n{'\n'.join(error_list)}")


except FileNotFoundError:
    print(f"Error: The file {file_path} was not found.")
    sys.exit(1)
except json.JSONDecodeError as e:
    print(f"Error: Failed to decode JSON from the file. Details: {e}")
    sys.exit(1)
except ValueError as e:
    print(e)
    sys.exit(1)